base/broadcast.jl

```diff
@@ -1211,4 +1211,29 @@ end
 end
 @inline broadcasted(::S, f, args...) where S<:BroadcastStyle = Broadcasted{S}(f, args)
 
+function _lazy end  # only used in `@:` macro; does not have to be callable
+# wrap the Broadcasted object in a tuple to avoid materializing
+@inline broadcasted(::typeof(_lazy), x) = (x,)
+
+"""
+    @: broadcasting_expression
+
+Construct a non-materialized broadcasted object from a `broadcasting_expression`
+and [`instantiate`](@ref) it.
+
+# Examples
+```jldoctest
+julia> bc = @: (1:3).^2;
+
+julia> bc isa Broadcast.Broadcasted  # it's not an `Array`
+true
+
+julia> sum(bc)  # summation without allocating an array
+14
+```
+"""
+macro (:)(ex)
+    return esc(:($instantiate(($_lazy.($ex))[1])))
+end
```

Comment:
Super cool syntax!

I like the fact that `instantiate` is called before passing the object on. If a non-macro variant gets exported, I think that should eagerly `instantiate` as well. That is, `@inline broadcasted(::typeof(_lazy), x) = (instantiate(x),)` instead of instantiating in the macro? Or am I missing some reason that this cannot work?

Reply:
Yea, that's possible.  Maybe better to do less in the macro.  Another way is to create a custom wrapper object and `instantiate` in `materialize`, like you did in the other PR.  But I'm seeing this as an implementation detail and it's better to use the simplest one.

Reply:
Unless `_lazy` gets exported in some form, I agree.